### ISSUE TITLE
aws s3store

### ISSUE THREAD

**Issue by rodentskie** [tus/tusd#1224]
**Question**
Is it possible to separate actual files from `.info`?

```
files/
info/
```

Setup works fine and as expected.

**Setup details**
Please provide following details, if applicable to your situation:
- Operating System: Ran it on AWS ECS Fargate
- Used tusd version: `"github.com/tus/tusd/v2/pkg/handler"` package I used on my go code.
- Used tusd data storage: AWS S3
- Used tus client library: uppy

**Comment by Acconut** [comment]
Yes, you can by configuring different `ObjectPrefix` and `ObjectMetadataPrefix` values: https://github.com/tus/tusd/blob/fb5b6209a7e1c807c9de8f081fa39ba8386b91d0/pkg/s3store/s3store.go#L108-L114

**Comment by rodentskie** [comment]
> Yes, you can by configuring different `ObjectPrefix` and `MetadataObjectPrefix` values:
> 
> https://github.com/tus/tusd/blob/fb5b6209a7e1c807c9de8f081fa39ba8386b91d0/pkg/s3store/s3store.go#L108-L114

Copy, will try to implement now and update in a while.

**Comment by rodentskie** [comment]
Works as expected. Here is my full `tusd` handler.

```
package tushandler

import (
	"context"
	"fmt"

	env "library/go/env/library/go"

	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"go.uber.org/zap"

	logger "library/go/logger/library/go"

	tusd "github.com/tus/tusd/v2/pkg/handler"
	"github.com/tus/tusd/v2/pkg/s3store"
)

func TusdHandler() *tusd.Handler {
	l := logger.LoggerConfig()
	zapLog, _ := l.Build()

	defer zapLog.Sync()

	region := env.GetEnv("REGION", "xxx")
	bucket := env.GetEnv("BUCKET", "xxx")
	accessKey := env.GetEnv("ACCESS_KEY_ID", "")
	secretKey := env.GetEnv("SECRET_ACCESS_KEY", "")

	cfg, err := config.LoadDefaultConfig(context.TODO(),
		config.WithRegion(region),
		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
			accessKey,
			secretKey,
			"", // Session token (optional)
		)),
	)

	if err != nil {
		zapLog.Fatal("error load aws creds",
			zap.Error(err),
		)
	}

	client := s3.NewFromConfig(cfg)
	store := s3store.New(bucket, client)
	composer := tusd.NewStoreComposer()

	store.ObjectPrefix = "files/"
	store.MetadataObjectPrefix = "info/"

	store.UseIn(composer)
	handler, err := tusd.NewHandler(tusd.Config{
		BasePath:                "/uploads/",
		StoreComposer:           composer,
		RespectForwardedHeaders: true,
	})
	if err != nil {
		panic(fmt.Errorf("unable to create handler: %s", err))
	}

	return handler
}

```

Next question would be the `Logger` of the `tusd.NewHandler`. Is it possible to use `zap` on this one?

**Comment by Acconut** [comment]
> Next question would be the `Logger` of the `tusd.NewHandler`. Is it possible to use `zap` on this one?

I don't use `zap`, but if it implements the `slog.Logger` interface, then yes. If not, you have to use or implement an adapter.

**Comment by rodentskie** [comment]
Hi @Acconut , I have a new feature to add.

The use-case is.

1. When we parse the JWT, there's a dynamic UUID as payload in there.
2. Same code above, the `store.ObjectPrefix = "files/"`, I wanna make it dynamic and get the uuid from the request header. Like into `store.ObjectPrefix = "files/<uuid here>". 

I'm blocked on how to get the `request` header into the function handler above.
I did try also `tusd.NewUnroutedHandler` but to no avail.
Any tip will be a big help, if it is not doable hearing it from you is totally fine also.

**Comment by Acconut** [comment]
You can use the PreUploadCallback in the [Config](https://pkg.go.dev/github.com/tus/tusd/v2@v2.6.0/pkg/handler#Config) struct to customize the upload ID. In the callback, the request information is also available.

**Comment by Acconut** [comment]
As mentioned in https://tus.github.io/tusd/advanced-topics/hooks/#hook-requests-and-responses, the upload ID is null for the pre-create hook because tusd didn't generate one at this stage yet. You can just generate a random ID on your own. Tusd's default method for ID generation is not publicly exported but you can copy it if you want to: https://github.com/tus/tusd/blob/main/internal/uid/uid.go
